### PR TITLE
Add AWS PubSub Support

### PR DESCRIPTION
**Pull request by DomBlack**
This commit adds support for AWS to Encore's PubSub system.
This is built using SNS topic's for publishing, and SQS queues
for each subscription. This allows topics to fan out to subscriptions
and for each subscription to manage it's own dead letter queue.

It relies on the ECS container or EC2 instance being granted
the correct roles to access the intended SNS topics and SQS queues.

### CODE REVIEW

File: runtime/pubsub/internal/aws/topic.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	_, err := t.sqsClient.GetQueueAttributes(t.ctx, &sqs.GetQueueAttributesInput{
+		QueueUrl: aws.String(implCfg.ProviderName),
+	})
+	if err != nil {
+		panic(fmt.Sprintf("unable to verify SQS queue attributes (may be missing IAM role allowing access): %v", err))
+	}
+
+	go func() {
+		for t.ctx.Err() == nil {
+			resp, err := t.sqsClient.ReceiveMessage(t.ctx, &sqs.ReceiveMessageInput{
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply runtime/pubsub/internal/aws/topic.go:63]: Yeah, which is why I put the QueueAttributes check above.

File: runtime/pubsub/internal/aws/topic.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+				QueueUrl:              aws.String(implCfg.ProviderName),
+				AttributeNames:        []sqsTypes.QueueAttributeName{"ApproximateReceiveCount"},
+				MaxNumberOfMessages:   1, // We only pull 1 message at a time, as the ackDeadline is per message
+				MessageAttributeNames: []string{"All"},
+				VisibilityTimeout:     int32(ackDeadline.Seconds()),
+				WaitTimeSeconds:       20, // Maximum allowed time
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply runtime/pubsub/internal/aws/topic.go:69]: If Long polling AWS only allows a configuration between 1 second and 20, so in this case it's the max that AWS supports.

File: runtime/pubsub/internal/aws/topic.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+			if err != nil && t.ctx.Err() == nil {
+				logger.Warn().Err(err).Msg("pubsub subscription failed, retrying in 5 seconds")
+				time.Sleep(5 * time.Second)
+				continue
+			}
+
+			for _, msg := range resp.Messages {
+				// Parse the message body
+				msgWrapper := &SNSMessageWrapper{}
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply runtime/pubsub/internal/aws/topic.go:80]: Yup, SNS wraps the message when the subscription forwards it to SQS

File: runtime/pubsub/internal/aws/topic.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+				// Get the delivery attempt number
+				deliveryAttempt, err := parseInt(msg.Attributes, "ApproximateReceiveCount")
+				if err != nil {
+					logger.Warn().Err(err).Str("msg_id", msgWrapper.MessageId).Msg("unable to parse receive count")
+				}
+
+				// Extract the attributes
+				attributes := make(map[string]string)
+				for key, value := range msgWrapper.MessageAttributes {
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply runtime/pubsub/internal/aws/topic.go:94]: The message attributes are not propergated to the SQS queue - that is unless we turn on `RawMessageDelivery` (which removes the wrapper).

However, I didn't want to remove the wrapper as it contains the publish time (to the topic), not the publish time to the queue (which could be a little bit of time later) - also the SQS message ID is different to the original SNS message ID, so by keeping the wrapper we preserve access to the original Message ID:


See: https://aws.amazon.com/blogs/compute/managing-amazon-sns-subscription-attributes-with-aws-cloudformation/
> When you set RawMessageDelivery to true, you get two outcomes. First, your message is delivered as is, with no metadata added. Second, your [message attributes](https://docs.aws.amazon.com/sns/latest/dg/SNSMessageAttributes.html) propagate from Amazon SNS to [Amazon SQS](https://aws.amazon.com/sqs/), when the subscribing endpoint is an Amazon SQS queue.

File: runtime/pubsub/internal/aws/topic.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+					case "String":
+						attributes[key] = value.Value
+					default:
+						logger.Warn().Err(err).Str("msg_id", msgWrapper.MessageId).Str("attr_name", key).Str("attr_type", value.Type).Msg("unsupported attribute data type")
+					}
+				}
+
+				// Call the callback, and if there was no error, then we can delete the message
+				err = f(t.ctx, msgWrapper.MessageId, msgWrapper.Timestamp, int(deliveryAttempt), attributes, []byte(msgWrapper.Message))
+				if err == nil {
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply runtime/pubsub/internal/aws/topic.go:105]: It's not, it's based on the `VisiblityTimeout` - so it does not backoff.

https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html

I guess in theory we could manually extract the "ApproximateDeliveryAttempt" and then calculate a backoff, then change the visibility of that one message, but that's a lot of API calls to do - and doesn't apply in the case the App panic's or crashes between reading the message and the modifying it's visibility timeout.

File: runtime/pubsub/internal/utils/utils.go [hunk 6]
```diff
@@ -216,7 +218,9 @@ func pointify(rval reflect.Value, ptrDepth int) reflect.Value {
 // configuration in the RetryPolicy
 func GetDelay(maxRetries int, minDelay, maxDelay time.Duration, attempt uint16) (shouldRetry bool, backoff time.Duration) {
 	if maxRetries == types.NoRetries || (int(attempt) > maxRetries && maxRetries != types.InfiniteRetries) {
-		return false, 0
+		// return the max delay here, so if we can't dead letter a message
+		// the could will at least apply the largest allowed backoff
+		return false, maxDelay
```
Comment (ekerfelt) [inline runtime/pubsub/internal/utils/utils.go:223]: hmmm. is this meant to be used if the deadlettering of a message fails? the original intent with the 0  was no retry.

**Review by ekerfelt — APPROVED** [review]

File: runtime/pubsub/internal/utils/utils.go (hunk 6 above)
Comment (DomBlack) [reply runtime/pubsub/internal/utils/utils.go:223]: In AWS we can't forcable deadletter something, so we just keep the backoff going until AWS decides to dead letter it

**Pull request merged**